src/modules/plus/filter_dynamicgpstext.c

```diff
@@ -0,0 +1,1714 @@
+			return mid;
+		}
+		if (gps_points[mid].time < video_time)
+			il = mid+1;
+		else
+			ir = mid-1;
+	}
+
+	//don't return the closest gps point if time difference is too large:
+	if (abs(video_time - gps_points[il].time) > MAX_GPS_DIFF) {
+	//	mlt_log_info(NULL, "filter_dynamicgps.c binary_search_gps: TIME DIFF too large (>MAX_GPS_DIFFs) return -1: vid_time(synced)%d, gps_time:%d", video_time, gps_points[il].time);
+		return -1;
+	}
+	pdata->last_searched_index = il;
+	return il;
+}
+
+/* Writes the smoothed version of gpx lat/lon + time(local) to a .gpx file
+ */
+static void _DEBUG_write_smoothed_gps_to_file(mlt_filter filter, int smooth_lvl)
+{
+	private_data* pdata = (private_data*)filter->child;
+	gps_point* gps_points = pdata->gps_points;
+	const int gps_points_size = pdata->gps_points_size;
+
+	int i;
+	char intro_gpx[] = "<?xml version=\"1.0\" encoding=\"UTF-8\" ?><gpx version=\"1.0\" creator=\"shotcut:gpstextoverlay test export\" xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\" xmlns=\"http://www.topografix.com/GPX/1/0\" xsi:schemaLocation=\"http://www.topografix.com/GPX/1/0 http://www.topografix.com/GPX/1/0/gpx.xsd\">\n"
+						"<trk><trkseg>\n";
+	char outro_gpx[] = "\n</trkseg></trk></gpx>";
+
+	FILE* f = NULL;
+	char fname[100];
+	sprintf(fname, "smoothed_gpx_smooth_%d.gpx", smooth_lvl);
+	f= fopen(fname, "w"); //created in home_dir by def
+	if (!f) {
+		mlt_log_warning(NULL, "filter_dynamicgps.c couldn't open file to write - %s", fname);
+		return;
+	}
+	fputs(intro_gpx, f);
+	for (i=0; i<gps_points_size; i++)
+	{
+		if (!has_valid_location_smoothed(gps_points[i]))
+			continue;
+
+		fprintf(f, "<trkpt lat=\"%f\" lon=\"%f\">", gps_points[i].lat_smoothed, gps_points[i].lon_smoothed);
+
+		if (gps_points[i].ele !=GPS_UNINIT)
+			fprintf(f, "<ele>%f</ele>", gps_points[i].ele);
+
+		char time[50] = {0};
+		seconds_to_timestring(gps_points[i].time, "%Y-%m-%dT%H:%M:%S.000Z", time);
+		fprintf(f, "<time>%s</time>", time);
+
+		if (gps_points[i].hr != GPS_UNINIT)
+			fprintf(f, "\n\t<extensions> <gpxtpx:TrackPointExtension> <gpxtpx:hr>%d</gpxtpx:hr> </gpxtpx:TrackPointExtension> </extensions>", gps_points[i].hr);
+		fprintf(f, "</trkpt>\n");
+	}
+	fputs(outro_gpx, f);
+	fclose(f);
+	mlt_log_info(NULL, "filter_dynamicgps.c done writing (%d points) smoothed file - %s", gps_points_size, fname);
+}
+
+/* Converts the bearing angle to a cardinal direction
+ * 8 divisions
+ */
+static char* bearing_to_compass(int x)
+{
+	if (x<=22.5 || x>=360-22.5)
+		return "N";
+	else if (x<45+22.5)
+		return "NE";
+	else if (x<=90+22.5)
+		return "E";
+	else if (x<90+45+22.5)
+		return "SE";
+	else if (x<=180+22.5)
+		return "S";
+	else if (x<180+45+22.5)
+		return "SW";
+	else if (x<=270+22.5)
+		return "W";
+	else if (x<270+45+22.5)
+		return "NW";
+
+	return "-";
+}
+
+/* Updates data in crt_point using lat/lon_smoothed;
+ * Only updates fields with calc_ prefix
+ */
+static void recalculate_gps_data(mlt_filter filter, int req_smooth)
```

Comment:
I wonder if you want to consider creating a separate file/class to encapsulate all the GPS processing. I could imagine in the future someone might make a producer that overlays GPS data on a map or something and they might want to share these GPS functions.

Reply:
Yes, I am now thinking of splitting the file into 3 : gps parser, gps processing and the core filter logic. It was easier to follow and debug/not bother about headers/visibility while writing it.